libnmstate/nm/device.py

```diff
@@ -50,12 +50,22 @@ def delete(dev):
 
 
 def reapply(dev, connection_profile):
-    """Reapply the given connection profile on the device."""
+    """
+    Reapply the given connection profile on the device.
+    In case the reapply operation fails, fallback to the connection profile
+    activation.
+    """
     mainloop = nmclient.mainloop()
     mainloop.push_action(_safe_reapply_async, dev, connection_profile)
 
 
 def _safe_reapply_async(dev, connection_profile):
+    # Special cases are known to exist (bugs) where reapply is not functioning
+    # correctly. Use activation instead. (https://bugzilla.redhat.com/1702657)
+    if _requires_activation(dev, connection_profile):
```

Comment:
Since the bug is already fixed upstream, it would make sense to only do the special handling if it is really necessary. Otherwise we cannot test against the master branch and check if it works there.

Reference: https://gitlab.freedesktop.org/NetworkManager/NetworkManager/commit/4ed72fa658c03790700ba9084e9328fe38afdee9

Reply:
The separate functions would not solve this problem AFAICS. The code would need to auto-detect based on NM's behavior whether or not to only try normal activation. If reapply does not fail directly then IMHO the correct solution would be to verify whether reapply worked and issue a normal activation if the outcome state after reapply is not the desired one. Not sure how feasible this is.

Reply:
Another idea to at least allow to disable this workaround when testing would be to check for an environment variable, for example `NMSTATE_DISABLE_WORKAROUND_RHBZ1702657` and skip the workaround if the variable is set.